src/strawman/collections/CollectionStrawMan6.scala

```diff
@@ -0,0 +1,980 @@
+
+  /** Iterator can be used only once */
+  trait IterableOnce[+A] {
+    def iterator: Iterator[A]
+  }
+
+  /** Base trait for instances that can construct a collection from an iterable */
+  trait FromIterable[+C[X] <: Iterable[X]] {
+    def fromIterable[B](it: Iterable[B]): C[B]
+  }
+
+  /** Base trait for companion objects of collections */
+  trait IterableFactory[+C[X] <: Iterable[X]] extends FromIterable[C] {
+    def empty[X]: C[X] = fromIterable(View.Empty)
+    def apply[A](xs: A*): C[A] = fromIterable(View.Elems(xs: _*))
+  }
+
+  /** Base trait for generic collections */
+  trait Iterable[+A] extends IterableOnce[A] with IterableLike[A, Iterable] {
+    /** The collection itself */
+    protected def coll: this.type = this
+ }
+
+  /** Base trait for sequence collections */
+  trait Seq[+A] extends Iterable[A] with SeqLike[A, Seq] {
+    def apply(i: Int): A
+    def length: Int
+  }
+
+  /** Base trait for linearly accessed sequences */
+  trait LinearSeq[+A] extends Seq[A] with SeqLike[A, LinearSeq] { self =>
+
+    /** To be overridden in implementations: */
+    def isEmpty: Boolean
+    def head: A
+    def tail: LinearSeq[A]
+
+    def iterator = new Iterator[A] {
+      private[this] var current: Seq[A] = self
+      def hasNext = !current.isEmpty
+      def next = { val r = current.head; current = current.tail; r }
+    }
+
+    def length: Int = iterator.length
+
+    @tailrec final def apply(n: Int): A =
+      if (n == 0) head else tail.apply(n - 1)
+
+    /** Optimized version of `drop` that avoids copying */
+    @tailrec final override def drop(n: Int) =
+      if (n <= 0) this else tail.drop(n - 1)
+  }
+
+  /** Base trait for strict collections that can be built using a builder.
+   *  @param  A    the element type of the collection
+   *  @param Repr  the type of the underlying collection
+   */
+  trait Buildable[+A, +Repr] extends Any with IterableMonoTransforms[A, Repr]  {
+
+    /** Creates a new builder. */
+    protected[this] def newBuilder: Builder[A, Repr]
+
+    /** Optimized, push-based version of `partition`. */
+    override def partition(p: A => Boolean): (Repr, Repr) = {
+      val l, r = newBuilder
+      coll.iterator.foreach(x => (if (p(x)) l else r) += x)
+      (l.result, r.result)
+    }
+
+    // one might also override other transforms here to avoid generating
+    // iterators if it helps efficiency.
+  }
+
+  /** Base trait for collection builders */
+  trait Builder[-A, +To] { self =>
+
+    /** Append an element */
+    def +=(x: A): this.type
+
+    /** Result collection consisting of all elements appended so far. */
+    def result: To
+
+    /** Bulk append. Can be overridden if specialized implementations are available. */
+    def ++=(xs: IterableOnce[A]): this.type = {
+      xs.iterator.foreach(+=)
+      this
+    }
+
+    /** A builder resulting from this builder my mapping the result using `f`. */
+    def mapResult[NewTo](f: To => NewTo) = new Builder[A, NewTo] {
```

Comment:
It looks like `ListBuffer` only uses copy-on-write because it is used internally as a builder. COW may add overhead because it has to check `aliased` on every append (but I wouldn't be surprised if the impact turns out to be negligible in practice due to branch prediction doing the right thing).